[DDP] Refactor uneven inputs to take GradBucket

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61020
* **#61019**

Changes uneven input logic of running allreduce to using `GradBucket` structure. This is to enable support for comm. hook with join in the next diff.

Differential Revision: [D29480027](https://our.internmc.facebook.com/intern/diff/D29480027/)

rohan-varma committed:
[DDP] Refactor uneven inputs to take GradBucket

Changes uneven input logic of running allreduce to using `GradBucket` structure. This is to enable support for comm. hook with join in the next diff.

Differential Revision: [D29480027](https://our.internmc.facebook.com/intern/diff/D29480027/)

[ghstack-poisoned]

SciPioneer commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -442,12 +442,19 @@ void Reducer::mark_variable_ready_sparse(size_t variable_index) {
 }
 
-std::vector<at::Tensor> Reducer::get_bucket_tensors() const {
+std::vector<c10d::GradBucket> Reducer::get_bucket_tensors() const {
```
Can we rename this function as well? Since we no longer directly return tensors, `get_grad_buckets` should be a better name.

SciPioneer commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -110,7 +110,7 @@ class TORCH_API Reducer {
       GradBucket grad_bucket);
 
   // Returns a vector of tensors in each bucket in sequential order.
```
Please update the comment accordingly.

SciPioneer commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -442,12 +442,19 @@ void Reducer::mark_variable_ready_sparse(size_t variable_index) {
+std::vector<c10d::GradBucket> Reducer::get_bucket_tensors() const {
   std::lock_guard<std::mutex> lock(mutex_);
-  std::vector<at::Tensor> bucketTensors;
+  std::vector<c10d::GradBucket> bucketTensors;
```
Same: rename it as "gradBuckets"

SciPioneer commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -442,12 +442,19 @@ void Reducer::mark_variable_ready_sparse(size_t variable_index) {
+    bucketTensors.emplace_back(
+      i,
+      // This API is only used for uneven inputs which requires zero tensors
+      at::zeros_like(buckets_[i].replicas[0].contents),
```
I think this can surprise users.  According to the function name, I won't expect any transformation after fetching the data.

What about moving ``zeros_like` to L1024 of file `distributed.py`?

rohan-varma replied:
Yeah, agreed it is not ideal. However we can't really do the `zeros_like` in python anymore because we cannot construct a new GradBucket and there's no API to change the tensor corresponding to a `GradBucket` (the latter can probably be misused).

If we really don't want this, we can add back being able to construct a `GradBucket` in python (which I removed recently), although this will require constructing an extra `GradBucket` when we really don't need to.

SciPioneer replied:
What about providing 3 functions and only expose 1 as a public API for handling uneven inputs?
1. get_grad_buckets w/ this flag
2. get_grad_buckets w/o this flag, which calls 1)
3. get_zeros_like_grad_buckets w/o this flag, which calls 1), and only expose this as a public API.

rohan-varma replied:
Updated based on this suggestion, PTAL

rohan-varma committed:
Update on "[DDP] Refactor uneven inputs to take GradBucket"


Changes uneven input logic of running allreduce to using `GradBucket` structure. This is to enable support for comm. hook with join in the next diff.

Differential Revision: [D29480027](https://our.internmc.facebook.com/intern/diff/D29480027/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor uneven inputs to take GradBucket"


Changes uneven input logic of running allreduce to using `GradBucket` structure. This is to enable support for comm. hook with join in the next diff.

Differential Revision: [D29480027](https://our.internmc.facebook.com/intern/diff/D29480027/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor uneven inputs to take GradBucket"


Changes uneven input logic of running allreduce to using `GradBucket` structure. This is to enable support for comm. hook with join in the next diff.

Differential Revision: [D29480027](https://our.internmc.facebook.com/intern/diff/D29480027/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor uneven inputs to take GradBucket"


Changes uneven input logic of running allreduce to using `GradBucket` structure. This is to enable support for comm. hook with join in the next diff.

Differential Revision: [D29480027](https://our.internmc.facebook.com/intern/diff/D29480027/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Refactor uneven inputs to take GradBucket"


Changes uneven input logic of running allreduce to using `GradBucket` structure. This is to enable support for comm. hook with join in the next diff.

Differential Revision: [D29480027](https://our.internmc.facebook.com/intern/diff/D29480027/)

[ghstack-poisoned]
